File: thesis-api/main/services/question_service/definition_processing_service.py
from __future__ import annotations
import textwrap
from services.question_service.edit_distance_service import EditDistanceService


class DefinitionProcessingService:
    def __init__(self, definition: str, answer: str, article_title: str):
        self.definition = definition
        self.answer = answer
        self.article_title = article_title
        self.line_length = 120
        self.definition_length = 300
        self.edit_distance_service = EditDistanceService()
        self.exception_word = ["the", "and", "are", "for", "not", "but", "had", "has", "was", "all", "any", "one",
                               "man", "out", "you", "his", "her", "can"]

    def standardize_definition_length(self) -> DefinitionProcessingService:
        definition_split = self.definition.split('.')
        current_definition_length = 0
        definition_array = []
        for sentence in definition_split:
            if current_definition_length + len(sentence) < self.definition_length or len(definition_array) == 0:
                definition_array.append(sentence)
                current_definition_length += len(sentence)
            else:
                break
        self.definition = ".".join(definition_array) + "."
        return self

    def wrap_text(self) -> DefinitionProcessingService:
        self.definition = textwrap.fill(self.definition, width=self.line_length)
        return self

    def remove_answer_from_definition(self) -> DefinitionProcessingService:
        text = self.definition
        if text[-2] == ".":
            text = text[:len(text)-1]
        delimiters_postfixes = [",", ".", "!", "?", ";", "]"]
        delimiters_prefixes = ["["]
        white_space_replacer = "###"
        text = text.replace("  ", " ")
        text = text.replace(" ", " " + white_space_replacer + " ")
        for delimiter in delimiters_postfixes:
            text = text.replace(delimiter, " " + delimiter)
        for delimiter in delimiters_postfixes:
            text = text.replace(delimiter, delimiter + " ")
        for delimiter in delimiters_postfixes:
            text = text.replace(delimiter + "  ", delimiter + "&&&")
        for delimiter in delimiters_postfixes:
            text = text.replace(delimiter + " ", delimiter + " " + white_space_replacer + " ")
        for delimiter in delimiters_postfixes:
            text = text.replace(delimiter + "&&&", delimiter + " ")
        for delimiter in delimiters_prefixes:
            text = text.replace(delimiter, delimiter + " ")
        for delimiter in delimiters_prefixes:
            text = text.replace(delimiter, " " + delimiter)
        for delimiter in delimiters_prefixes:
            text = text.replace("  " + delimiter, "&&&" + delimiter)
        for delimiter in delimiters_prefixes:
            text = text.replace(" " + delimiter, " " + white_space_replacer + " " + delimiter)
        for delimiter in delimiters_prefixes:
            text = text.replace("&&&" + delimiter, " " + delimiter)

        words = text.split()
        distances_to_right_answer = {}
        for word in words:
            if word not in delimiters_prefixes and word not in delimiters_postfixes and word != white_space_replacer:
                if len(word) > 2 and word not in self.exception_word:
                    for answer_word in self.answer.split():
                        if word not in distances_to_right_answer:
                            distances_to_right_answer[word] = self.edit_distance_service.edit_distance(word,
                                                                                                       answer_word)
                        else:
                            distances_to_right_answer[word] = min(distances_to_right_answer[word],
                                                                  self.edit_distance_service.edit_distance(word,
                                                                                                           answer_word))

        if self.answer != self.article_title:
            for word in words:
                if len(word) > 2 and word not in self.exception_word:
                    if word not in delimiters_prefixes and word not in delimiters_postfixes \
                            and word != white_space_replacer:
                        for answer_word in self.article_title.split():
                            distances_to_right_answer[word] = min(distances_to_right_answer[word],
                                                                  self.edit_distance_service.edit_distance(word,
                                                                                                           answer_word))

        distances_values_keys = [k for k, v in sorted(distances_to_right_answer.items(), key=lambda item: item[1])]
        distances_values_keys = distances_values_keys[:max(len(distances_to_right_answer.keys())//20,
                                                           max(len(self.answer.split()),
                                                               len(self.article_title.split())))]
        summary_censored = []
        for word in words:
            if word not in delimiters_postfixes and word not in delimiters_prefixes and word != white_space_replacer:
                if word not in distances_values_keys:
                    if (word.startswith(self.answer) or self.answer.startswith(word) or \
                            word.startswith(self.article_title) or self.article_title.startswith(word)) \
                            and len(word) > 2 and word not in self.exception_word:
                        summary_censored.append("____")
                    elif word in distances_to_right_answer and distances_to_right_answer[word] <= 1:
                        summary_censored.append("____")
                    else:
                        summary_censored.append(word)
                else:
                    summary_censored.append("____")
            else:
                summary_censored.append(word)
        self.definition = "".join(summary_censored).replace(white_space_replacer, " ")
        return self

    def get_definition(self) -> str:
        return self.definition

File: thesis-api-v2/src/main/java/pl/edu/agh/quizzesthesis/business/DefinitionProcessingService.java
package pl.edu.agh.quizzesthesis.business;

import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;
import pl.edu.agh.quizzesthesis.api.dto.QuestionResponse;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static java.lang.Math.max;
import static java.lang.Math.min;

@Service
@AllArgsConstructor
public class DefinitionProcessingService {

    private static final Set<String> EXCEPTIONAL_WORDS = Set.of(
            "the", "and", "are", "for", "not", "but", "had", "has", "was", "all", "any", "one", "man", "out", "you",
            "his", "her", "can"
    );
    private static final Set<String> DELIMITER_POSTFIXES = Set.of(",", ".", "!", "?", ")", ";", "]");
    private static final Set<String> DELIMITER_PREFIXES = Set.of("(", "[");
    private static final String WHITE_SPACE_REPLACER = "###";
    private static final int LINE_LENGTH = 120;
    private static final int DEFINITION_LENGTH = 300;

    private final EditDistanceService editDistanceService;

    public DefinitionProcessing startProcessing(String definition, String answer, String articleTitle) {
        return new DefinitionProcessing(definition, answer, articleTitle);
    }

    public class DefinitionProcessing {

        private final String definition;
        private final String answer;
        private final String articleTitle;

        private DefinitionProcessing(String definition, String answer, String articleTitle) {
            this.definition = definition;
            this.answer = answer;
            this.articleTitle = articleTitle;
        }

        public String getDefinition() {
            return definition;
        }

        public DefinitionProcessing standardizeDefinitionLength() {
            String[] definitionSplit = definition.split("\\.");
            int currentDefinitionLength = 0;
            var definitionList = new ArrayList<String>();
            for (String sentence : definitionSplit) {
                if (currentDefinitionLength + sentence.length() < DEFINITION_LENGTH || definitionList.size() == 0) {
                    definitionList.add(sentence);
                    currentDefinitionLength += sentence.length();
                } else {
                    break;
                }
            }
            return new DefinitionProcessing(String.join(".", definitionList) + ".", answer, articleTitle);
        }

        public DefinitionProcessing removeAnswerFromDefinition() {
            String text = definition;
            if (text.charAt(text.length() - 2) == '.') {
                text = text.substring(0, text.length() - 1);
            }
            text = text.replace("  ", " ");
            text = text.replace(" ", " " + WHITE_SPACE_REPLACER + " ");

            for (String delimiter : DELIMITER_POSTFIXES) {
                text = text.replace(delimiter, " " + delimiter);
            }
            for (String delimiter : DELIMITER_POSTFIXES) {
                text = text.replace(delimiter, delimiter + " ");
            }
            for (String delimiter : DELIMITER_POSTFIXES) {
                text = text.replace(delimiter + "  ", delimiter + "&&&");
            }
            for (String delimiter : DELIMITER_POSTFIXES) {
                text = text.replace(delimiter + " ", delimiter + " " + WHITE_SPACE_REPLACER + " ");
            }
            for (String delimiter : DELIMITER_POSTFIXES) {
                text = text.replace(delimiter + "&&&", delimiter + " ");
            }
            for (String delimiter : DELIMITER_PREFIXES) {
                text = text.replace(delimiter, delimiter + " ");
            }
            for (String delimiter : DELIMITER_PREFIXES) {
                text = text.replace(delimiter, " " + delimiter);
            }
            for (String delimiter : DELIMITER_PREFIXES) {
                text = text.replace("  " + delimiter, "&&&" + delimiter);
            }
            for (String delimiter : DELIMITER_PREFIXES) {
                text = text.replace(" " + delimiter, " " + WHITE_SPACE_REPLACER + " " + delimiter);
            }
            for (String delimiter : DELIMITER_PREFIXES) {
                text = text.replace("&&&" + delimiter, " " + delimiter);
            }

            String[] words = Arrays.stream(text.split("\s"))
                    .toArray(String[]::new);

            var distancesToRightAnswer = new HashMap<String, Float>();
            for (String word : words) {
                if (!DELIMITER_PREFIXES.contains(word) && !DELIMITER_POSTFIXES.contains(word) && !word.equals(WHITE_SPACE_REPLACER)) {
                    if (word.length() > 2 && !EXCEPTIONAL_WORDS.contains(word)) {
                        for (String answerWord : answer.split("\s")) {
                            if (!distancesToRightAnswer.containsKey(word)) {
                                distancesToRightAnswer.put(word, editDistanceService.editDistance(word, answerWord));
                            } else {
                                distancesToRightAnswer.put(
                                        word,
                                        min(
                                                distancesToRightAnswer.get(word),
                                                editDistanceService.editDistance(word, answerWord)
                                        )
                                );
                            }
                        }
                    }
                }
            }

            if (!answer.equalsIgnoreCase(articleTitle)) {
                for (String word : words) {
                    if (word.length() > 2 && !EXCEPTIONAL_WORDS.contains(word)) {
                        if (!DELIMITER_PREFIXES.contains(word) && !DELIMITER_POSTFIXES.contains(word) && !word.equals(WHITE_SPACE_REPLACER)) {
                            for (String answerInTitle : articleTitle.split("\s")) {
                                distancesToRightAnswer.put(
                                        word,
                                        min(
                                                distancesToRightAnswer.get(word),
                                                editDistanceService.editDistance(word, answerInTitle)
                                        )
                                );
                            }
                        }
                    }
                }
            }

           var distanceValueKeys = distancesToRightAnswer.entrySet().stream()
                    .sorted((a, b) -> Float.compare(a.getValue(), b.getValue()))
                    .limit(limitDistanceValueKeys(distancesToRightAnswer))
                    .map(Map.Entry::getKey)
                    .toList();

            var summaryCensored = new ArrayList<String>();
            for (String word : words) {
                if (!DELIMITER_POSTFIXES.contains(word) && !DELIMITER_PREFIXES.contains(word) && !word.equals(WHITE_SPACE_REPLACER)) {
                    if (!distanceValueKeys.contains(word)) {
                        if (word.startsWith(answer) || answer.startsWith(word) ||
                                word.startsWith(articleTitle) || articleTitle.startsWith(word) &&
                                word.length() > 2 && !EXCEPTIONAL_WORDS.contains(word)) {
                            summaryCensored.add("____");
                        } else if (distancesToRightAnswer.containsKey(word) && distancesToRightAnswer.get(word) <= 1) {
                            summaryCensored.add("____");
                        } else {
                            summaryCensored.add(word);
                        }
                    } else {
                        summaryCensored.add("____");
                    }
                } else {
                    summaryCensored.add(word);
                }
            }
            String newDefinition = String.join("", summaryCensored).replace(WHITE_SPACE_REPLACER, " ");
            return new DefinitionProcessing(newDefinition, answer, articleTitle);
        }

        private int limitDistanceValueKeys(Map<String, Float> distancesToRightAnswer) {
            return max(
                    distancesToRightAnswer.size() / 20,
                    max(answer.split("\s").length, articleTitle.split("\s").length)
            );
        }
    }
}
